internal/security/fileprovider/tokenconfig.go

```diff
@@ -80,3 +88,77 @@ func LoadTokenConfig(fileOpener fileioperformer.FileIoPerformer, path string, to
 
 	return nil
 }
+
+// GetTokenConfigFromEnv function gets a list of token service keys
+// from environment variable and populates the default configuration with
+// default token parameters and policies
+// the function returns a TokenConfFile map instance and error if any
+// if the environment variable is not present or the value of that is empty
+// then it will return empty map
+// if the value for the list is not well-formed, not comma-separated
+// then it will return an error
+func GetTokenConfigFromEnv() (TokenConfFile, error) {
+	emptyTokenConfig := make(TokenConfFile)
+
+	addTokenList := os.Getenv(addSecretstoreTokensEnvKey)
+	if strings.TrimSpace(addTokenList) == "" {
+		return emptyTokenConfig, nil
+	}
+
+	// the list of service names is comma-separated
+	tokenConfigFromEnv := make(TokenConfFile)
+	serviceNameList := strings.Split(addTokenList, ",")
+	// regex for valid service name as key
+	// the service name eventually becomes part of the URL to Vault's API call
+	// / is not ok and backslack \ is not ok comma , is not ok
+	// and the the length of name is upto 512 characters
+	serviceNameRegx := regexp.MustCompile(`^[^.*\/\\\,]*[^\W]*[\w.]{1,512}$`)
+
+	for _, name := range serviceNameList {
+		serviceName := strings.TrimSpace(name)
+		if serviceName == "" {
+			return emptyTokenConfig, errors.New("cannot have empty service name as key from environment variable")
+		}
+
+		if !serviceNameRegx.MatchString(serviceName) {
+			return emptyTokenConfig, fmt.Errorf("Invalid service name: %s as key from environment variable", serviceName)
+		}
+
+		// with default service configuration
+		tokenConfigFromEnv[serviceName] = ServiceKey{
+			UseDefaults: true,
+		}
+	}
+
+	return tokenConfigFromEnv, nil
+}
+
+// mergeWith function takes another TokenConfFile and merges with the current one
+// to return a new TokenConfFile.
+// The merging is based on the key of TokenConfFile's map
+// if the key of another has already been existing on the current one,
+// then the former will replaces the latter.
+func (tf TokenConfFile) mergeWith(another TokenConfFile) TokenConfFile {
+	if len(another) == 0 {
+		// nothing to be merged, return itself
+		return tf
+	}
+
```

Comment:
could do that and maybe that way is more efficient; functionality wise the current implementation is already taken care by the following few lines of code when `tf` is empty, though.

Reply:
It isn't about efficiency. The way you are using this func, `tf`=from env and `another`=from config. `another` will never be empty since loaded from config, but `tf` could be.  All I am suggesting it to add a check of `tf` for empty like your are doing for `another` to cover if either are empty.

Reply:
This function runs once per boot.  Is it really necessary to optimize it?

Reply:
it is just a simple check to do so I think we should do it.

Reply:
It isn't about optimizing it... ;-)